kalatori-ah/src/rpc.rs

```diff
@@ -0,0 +1,1352 @@
+use crate::{
+    database::{Database, Invoice, InvoiceStatus, ReadInvoices},
+    shutdown, Account, Balance, BlockNumber, Decimals, Hash, Nonce, OnlineClient, RuntimeConfig,
+    Usd, EXPECTED_USDX_FEE, SCANNER_TO_LISTENER_SWITCH_POINT,
+};
+use anyhow::{Context, Result};
+use reconnecting_jsonrpsee_ws_client::ClientBuilder;
+use serde::{Deserialize, Deserializer};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    error::Error,
+    fmt::{self, Arguments, Display, Formatter, Write},
+    sync::Arc,
+};
+use subxt::{
+    backend::{
+        legacy::{LegacyBackend, LegacyRpcMethods},
+        rpc::{RpcClient, RpcSubscription},
+        Backend, BackendExt, RuntimeVersion,
+    },
+    blocks::{Block, BlocksClient},
+    config::{DefaultExtrinsicParamsBuilder, Header},
+    constants::ConstantsClient,
+    dynamic::{self, Value},
+    error::RpcError,
+    ext::{
+        futures::TryFutureExt,
+        scale_decode::DecodeAsType,
+        scale_value::{self, At},
+        sp_core::{
+            crypto::{AccountId32, Ss58AddressFormat},
+            sr25519::Pair,
+        },
+    },
+    storage::{Storage, StorageClient},
+    tx::{PairSigner, SubmittableExtrinsic, TxClient},
+    Config, Metadata,
+};
+use tokio::sync::{mpsc::UnboundedSender, RwLock};
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+
+pub const MODULE: &str = module_path!();
+
+const MAX_BLOCK_NUMBER_ERROR: &str = "block number type overflow is occurred";
+const BLOCK_NONCE_ERROR: &str = "failed to fetch an account nonce by the scanner client";
+
+// Pallets
+
+const SYSTEM: &str = "System";
+const UTILITY: &str = "Utility";
+const ASSETS: &str = "Assets";
+
+async fn fetch_best_block(methods: &LegacyRpcMethods<RuntimeConfig>) -> Result<Hash> {
+    methods
+        .chain_get_block_hash(None)
+        .await
+        .context("failed to get the best block hash")?
+        .context("received nothing after requesting the best block hash")
+}
+
+async fn fetch_runtime(
+    methods: &LegacyRpcMethods<RuntimeConfig>,
+    backend: &impl Backend<RuntimeConfig>,
+) -> Result<(Metadata, RuntimeVersion)> {
+    let best_block = fetch_best_block(methods).await?;
+
+    Ok((
+        fetch_metadata(backend, best_block)
```

Comment:
shouldn't this be finalized block?

Reply:
`fetch_runtime` is used only with the API client that requires to be in sync with the best chain, not the finalized one.